arangod/RestHandler/RestTasksHandler.cpp

```diff
@@ -0,0 +1,341 @@
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Dan Larkin-York
+////////////////////////////////////////////////////////////////////////////////
+
+#include "RestTasksHandler.h"
+
+#include "ApplicationFeatures/ApplicationServer.h"
+#include "Basics/VelocyPackHelper.h"
+#include "Cluster/ServerState.h"
+#include "Rest/HttpRequest.h"
+#include "Scheduler/Scheduler.h"
+#include "Scheduler/SchedulerFeature.h"
+#include "V8/v8-globals.h"
+#include "V8/v8-vpack.h"
+#include "V8Server/V8DealerFeature.h"
+#include "VocBase/Methods/Tasks.h"
+
+#include <velocypack/Builder.h>
+#include <velocypack/velocypack-aliases.h>
+
+using namespace arangodb::basics;
+using namespace arangodb::rest;
+
+namespace arangodb {
+
+RestTasksHandler::RestTasksHandler(GeneralRequest* request,
+                                   GeneralResponse* response)
+    : RestVocbaseBaseHandler(request, response) {}
+
+RestStatus RestTasksHandler::execute() {
+  auto const type = _request->requestType();
+
+  switch (type) {
+    case rest::RequestType::POST: {
+      registerTask(false);
+      break;
+    }
+    case rest::RequestType::PUT: {
+      registerTask(true);
+      break;
+    }
+    case rest::RequestType::DELETE_REQ: {
+      deleteTask();
+      break;
+    }
+    case rest::RequestType::GET: {
+      getTasks();
+      break;
+    }
+    default: {
+      generateError(rest::ResponseCode::METHOD_NOT_ALLOWED,
+                    TRI_ERROR_HTTP_METHOD_NOT_ALLOWED);
+    }
+  }
+  return RestStatus::DONE;
+}
+
+/// @brief returns the short id of the server which should handle this request
+uint32_t RestTasksHandler::forwardingTarget() {
+  rest::RequestType const type = _request->requestType();
+  if (type != rest::RequestType::POST &&
+      type != rest::RequestType::PUT &&
+      type != rest::RequestType::GET &&
+      type != rest::RequestType::DELETE_REQ) {
+    return false;
+  }
+
+  std::vector<std::string> const& suffixes = _request->suffixes();
+  if (suffixes.size() < 1) {
+    return false;
+  }
+
+  uint64_t tick = arangodb::basics::StringUtils::uint64(suffixes[0]);
+  uint32_t sourceServer = TRI_ExtractServerIdFromTick(tick);
+
+  return (sourceServer == ServerState::instance()->getShortId())
+      ? 0
+      : sourceServer;
+}
+
+void RestTasksHandler::getTasks() {
+  std::vector<std::string> const& suffixes = _request->decodedSuffixes();
+  if (suffixes.size() > 1) {
+    generateError(rest::ResponseCode::BAD, TRI_ERROR_HTTP_SUPERFLUOUS_SUFFICES,
+                  "superfluous parameter, expecting /_api/tasks[/<id>]");
+    return;
+  }
+
+  ExecContext const* exec = ExecContext::CURRENT;
+  if (exec != nullptr) {
+    if (exec->databaseAuthLevel() != auth::Level::RW) {
+      generateError(rest::ResponseCode::FORBIDDEN, TRI_ERROR_FORBIDDEN,
+                    "unregister task needs db RW permissions");
+      return;
+    } else if (!exec->isSuperuser() && ServerState::readOnly()) {
```

Comment:
read-only mode should not block fetching tasks IMHO, as fetching tasks only reads them but does not modify.